README.md
# Kodi Mirror
A Kodi plugin that allows you to mirror the content from another Kodi installation.

## Setup
1. On the server Kodi installation, allow control access via HTTP and 'Remote Control'
	- Settings -> Services -> Webserver -> [Check] Allow access of XBMC/Kodi via HTTP (Necessary)
	- Settings -> Services -> Remote control -> [Check] Allow programs on other systems to control XBMC/Kodi (Recommended)
2. Note down the server IP and webserver port.
3. Install the Kodi Mirror plugin on the client installations.
4. Click 'Add Server' within the plugin and input the IP and port as noted above.
5. Watch

## Settings
- Timeout: How long to try and get the playing content from the server.
- Max Latency: The maximum time difference of the currently playing media that is allowed. (This would need to be larger the slower your network is)

### Remote Access
In theory, this should allow for remote mirror access. Two ports will need to be forwarded to the server: 9090 and the webserver port.
9090 is the default port for the TCP Socket API, this can be changed via advanced settings, but currently this plugin does not support it.
This option has not been tested, if you happen to try it out, let me know how it works out: wilingua@gmail.com

## Author
James Swindle
wilingua@gmail.com


plugin.video.mirror/changelog.txt
v0.1
-Initial Development
-Got stuff working in a decent manor

v0.2
-Added latency and timeout settings

v0.3
-Fixed bug of 'opening stream' window persisting

v0.3.1
-Removed rogue import which hopefully fixes issue on OSX

plugin.video.mirror/addon.py
import urllib
import urlparse
import operator
import time, random

# Dharma compatibility (import md5)
try:
    import hashlib
except:
    import md5

import xbmc
import xbmcaddon
import xbmcgui
import xbmcplugin

import serverSettings
from MirrorPlayer import MirrorPlayer


base_url = sys.argv[0]
addon_handle = int(sys.argv[1])
args = urlparse.parse_qs(sys.argv[2][1:])

myaddon = xbmcaddon.Addon()
path = xbmc.translatePath(myaddon.getAddonInfo('profile')).decode("utf-8")

def _build_url(query):
    return base_url + '?' + urllib.urlencode(query)
def _translation(id):
    return myaddon.getLocalizedString(id).encode('utf-8')
def _getParam(name, isSingle = True):
    val = args.get(name, None)
    if val is not None and isSingle:
        val = val[0]
    return val
def _get_SID():
    t1 = time.time()
    t2 = t1 + random.getrandbits(32)
    try: 
        # Eden & + compatible
        base = hashlib.md5( str(t1 +t2) )
    except:
        # Dharma compatible
        base = md5.new( str(t1 +t2) )
    sid = base.hexdigest()
    return sid
    
    
def main():
    servers = serverSettings.readServers()
    for server in servers:
        li = xbmcgui.ListItem(unicode(server['name']))
        
        runner = xbmc.translatePath(myaddon.getAddonInfo('path')).decode("utf-8") + '/contextHandler.py'
        arg = 'mode=remove&id=' + str(server['id'])
        li.addContextMenuItems([(_translation(30001), "XBMC.RunScript(" + runner + ", " + arg + ")")])
        li.setProperty('IsPlayable', 'true')
        xbmcplugin.addDirectoryItem(handle=addon_handle, url=_build_url({'mode': 'server', 'id': str(server['id'])}), listitem=li)
        
    #Create server    
    li = xbmcgui.ListItem(_translation(30002))
    xbmcplugin.addDirectoryItem(handle=addon_handle, url=_build_url({'mode': 'add'}), listitem=li, isFolder=True)
    
    xbmcplugin.addSortMethod(addon_handle, xbmcplugin.SORT_METHOD_TITLE)
    xbmcplugin.endOfDirectory(addon_handle)
    
def addServer():
    dialog = xbmcgui.Dialog()
    keyboard = xbmc.Keyboard("", _translation(32001))
    keyboard.doModal()
    if not keyboard.isConfirmed():
        return False
        
    ip = dialog.input(_translation(32002), "", xbmcgui.INPUT_IPADDRESS)
    if ip is None or ip == "":
        return False
    
    port = dialog.input(_translation(32003), "8080", xbmcgui.INPUT_NUMERIC)
    if port is None or port == "":
        return False
    
    serverdata = {"name": keyboard.getText(), "host": ip, "port": int(port), "id": _get_SID()}
    serverSettings.addServer(serverdata)
    xbmc.executebuiltin("Container.Refresh")
    return True

def startMirror(id):
    server = serverSettings.getServer(id)
    if server is None:
        xbmc.log('kodi-mirror::No server for id: ' + id, xbmc.LOGERROR)
        xbmcgui.Dialog().ok('Error', 'That server doesn\'t exists')
        return False
    player = MirrorPlayer()
    player.withSocket(server['host'], int(server['port']))
    listitem = player.playFromServer()
    if listitem is not None:
        xbmcplugin.setResolvedUrl(addon_handle, True, listitem)
    else:
        xbmcplugin.setResolvedUrl(addon_handle, False, xbmcgui.ListItem(unicode(server['name'])))
    while not xbmc.abortRequested and player.isActive():
        xbmc.sleep(500) 
    del player
    
mode = _getParam('mode')
if mode is None:
    main()
elif mode == 'add':
    addServer()
elif mode == 'server':
    startMirror(_getParam('id'))
                
        

plugin.video.mirror/RPCSocket.py
import sys
import socket
import threading
import xbmcaddon
try:
    import simplejson as json
except:
    import json
class RPCSocket:
    def __init__(self, host, port, callbacks=None):
        if host is not None and port is not None:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self._callbacks = callbacks
            self._keepAlive = True
            self._connect(host, port)

    def _connect(self, host, port):
        timeout = abs(float(xbmcaddon.Addon().getSetting('timeout'))) * 1000
        self.sock.settimeout(timeout)
        self.sock.connect((host, port))
        self.sock.settimeout(None)
        threading.Thread(target=self._wait).start()    
            
    def setCallback(self, method, callback):
        if self._callbacks is None:
            self._callbacks = {}
        self._callbacks[str(method)] = callback
        
    def _wait(self):
        datas = []
        data = ""
        while self._keepAlive:
            try:
                a = self.sock.recv(1)
            except:
                break
            if not a:
                break
            data += a
            try:
                js = json.loads(data)
                data = ""
            except:
                continue
            if self._callbacks is not None and js:
                if 'id' in js and str(js['id']) in self._callbacks:
                    self._callbacks[str(js['id'])](self, js)
                elif 'method' in js and js['method'] in self._callbacks:
                    self._callbacks[js['method']](self, js)
                    
    def isAlive(self):
        return self._keepAlive

    def send(self, payload, callback=None):
            if callback is not None:
                self.setCallback(payload['id'], callback)
            sent = self.sock.send(json.dumps(payload))
            if sent == 0:
                raise RuntimeError("socket connection broken")
    def kill(self):
        try:
            self.sock.shutdown(1)
            self.sock.close()
        finally:
            self._keepAlive = False

plugin.video.mirror/serverSettings.py
try:
    import simplejson as json
except:
    import json
import os
from os.path import basename
import shutil
import time
from urlparse import urlparse

import requests

import xbmc
import xbmcaddon


_path = xbmc.translatePath(xbmcaddon.Addon('plugin.video.mirror').getAddonInfo('profile')).decode("utf-8")

if not os.path.exists(_path):
    os.makedirs(_path)
    
_serversFile = _path + '/servers.json'
    
def readServers():
    try:
        file = open(_serversFile, 'r')
        data = json.load(file)
        file.close()
        return data
    except:
        return []

def addServer(server):
    servers = readServers()
    servers.append(server)
    file = open(_serversFile, 'w')
    json.dump(servers, file)
    file.close()

def removeServer(id):
    servers = readServers()
    newList = []
    for server in servers:
        if 'id' in server and server['id'] != id:
            newList.append(server)
    
    file = open(_serversFile, 'w')
    json.dump(newList, file)
    file.close()    

def getServer(id):
    servers = readServers()
    for server in servers:
        if 'id' in server and server['id'] == id:
            return server    
    return None

plugin.video.mirror/MirrorPlayer.py
import xbmc, xbmcgui, xbmcaddon, xbmcplugin, xbmcvfs

import requests,time
try:
    import simplejson as json
except:
    import json
from RPCSocket import RPCSocket

musicprops = ["file", "duration", "year", "genre", "album", 
              "artist", "title", "rating", "lyrics", 
              "playcount", "lastplayed"]
pictureprops = ["title", "file"]
videoprops = ["title", "file", "genre", "year", "episode", "season",
              "top250", "rating", "playcount", "director",
              "mpaa", "plot", "plotoutline", "originaltitle",
              "duration", "studio", "tagline", "writer", "premiered",
              "lastplayed", "album", "votes", "trailer"]

class MirrorPlayer( xbmc.Player ):

    def __init__(self, *args, **kwargs):
        self._totalTime = 999999
        self._lastPos = 0
        self._pid = None
        self._type = ""
        self._webserv = ""
        self._itemId = -1 
        self._playstate = -1
        self._isActive = True
        super(MirrorPlayer, self).__init__(*args, **kwargs)
        
    def withSocket(self, host, port):
        self._webserv = 'http://%s:%d/' % (str(host), int(port))
        try:
            self.socket = RPCSocket(host, 9090, {'Player.OnPlay': self._onPlay, 'Player.OnPause': self._onPause, 'Player.OnStop': self._onStop})
        except:
            xbmcgui.Dialog().ok('Error', 'Unable to connect to the server via TCP', 'Please allow programs on other systems to control XBMC', 'until then, reduced functionality is available')
            self.socket = None
            return False
        return True
    
    #TODO detect speed changes from server
    def _onPlay(self, socket, data):
        xbmc.log('kodi-mirror::onPlay JSON: ' + json.dumps(data), xbmc.LOGDEBUG)
        if not self.isPaused():
            self._playstate = 1
        if self._playstate == 1:
            return False
        if data is not None:
            id = data['params']['data']['item']['id']
        else:
            id = None
        if id == None or int(self._itemId) == int(id):
            self.pause()
        else:
            self.playFromServer()
        
    def _onPause(self, socket, data):
        xbmc.log('kodi-mirror::onPause JSON: ' + json.dumps(data), xbmc.LOGDEBUG)
        if self.isPaused():
            self._playstate = 0
        if self._playstate != 0:
            self.pause()
    
    def _onStop(self, socket, data):
        xbmc.log('kodi-mirror::onStop JSON: ' + json.dumps(data), xbmc.LOGDEBUG)
        self._itemId = -1
        self._playstate = -1
        xbmc.executebuiltin("PlayerControl(Stop)")
        if self.socket is None or self.socket.isAlive():
            self._waitForItem()
        
    def isActive(self):
        return self._isActive
            
    def onPlayBackStopped(self):
        xbmc.log('kodi-mirror::Player Stopped', xbmc.LOGDEBUG)
        if self._playstate == -1:
            return
        if self.socket is not None:
            self.socket.kill()
            xbmc.log('kodi-mirror::Socket Killed', xbmc.LOGDEBUG)
        self._playstate = 0
        self._isActive = False
        
    def playFromServer(self):
        self._playstate = 1
        current = self._getItem()
        if 'error' in current:
            xbmc.log('kodi-mirror::' + current['error'], xbmc.LOGERROR)
            self._itemId = -1
            self._playstate = -1
            return self._waitForItem()
        elif current['id'] != self._itemId:
            super(MirrorPlayer, self).play(current['stream'], current['listItem'])
            self._itemId = current['id']  
            return current['listItem']
        
    def onPlayBackStarted( self ):
        #xbmc.executebuiltin('Dialog.Close(all, true)')
        #time.sleep(2)
        self._totalTime = self.getTotalTime()
        self.checkLatency()
        
    def checkLatency(self):
        latency = abs(int(xbmcaddon.Addon().getSetting('max_latency')))
        
        while not xbmc.abortRequested and self._playstate == 1 and self.isPlaying():
            seek = self._getSeek()
            if abs(seek['seek'] - self.getTime()) > latency:
                self.seekTime(seek['seek'])
                if seek['speed'] == 0:
                    self._onPause(None, None)
                else:
                    self._onPlay(None, None)
            xbmc.sleep(latency * 1000) 
        
    def onPlayBackResumed(self):
        self._playstate = 1
        self.checkLatency()
    
    def onPlayBackPaused(self):
        self._playstate = 0

    def onPlayBackEnded( self ):
        self._playstate = -1 
        if self.socket is None or self.socket.isAlive():
            self._waitForItem()
            
    def isPaused(self):
        start_time = self.getTime()
        time.sleep(1)
        if self.getTime() != start_time:
            return False
        else:
            return True
        
    def _waitForItem(self):
        if self._playstate != -1:
            return False
        self._pid = None
        self._itemId = -1
        dialog = xbmcgui.DialogProgress()
        dialog.create('Server Status', 'Waiting for content to be played...')
        dialog.update(0)
        count = 0
        timeout = abs(float(xbmcaddon.Addon().getSetting('timeout'))) * 2
        while self._playstate == -1 and self._isActive:
            item = self._getItem()
            if count > timeout or dialog.iscanceled():
                self._isActive = False
                break
            elif 'error' not in item:
                dialog.close()
                return self.playFromServer()
            else:
                upd = int((count / timeout) * 100)
                dialog.update(upd)
                xbmc.sleep(500)
                count += 1
        dialog.close()
        if not self._isActive:
            self.socket.kill()
        return None
                    
    def _getItem(self):
        if self._pid is None:
            payload = {"jsonrpc": "2.0", "method": "Player.GetActivePlayers", "id": 1}
            r = requests.post(self._webserv + 'jsonrpc', data=json.dumps(payload))
            d = r.json()
            xbmc.log('kodi-mirror::activePlayers JSON: ' + json.dumps(d), xbmc.LOGDEBUG)
            if 'error' in d:
                return {'error': d['error']}
            if len(d['result']) < 1:
                return {'error': 'No Item Playing'}
            self._type = d['result'][0]['type']
            self._pid = d['result'][0]['playerid']
        
        payload = {"jsonrpc": "2.0", "method": "Player.GetItem", "params": {"properties": videoprops, "playerid": self._pid}, "id": "getItem"}
        r = requests.post(self._webserv + 'jsonrpc', data=json.dumps(payload))
        d = r.json()
        xbmc.log('kodi-mirror::getItem JSON: ' + json.dumps(d), xbmc.LOGDEBUG)
        if 'error' in d:
            return {'error': d['error']}
        try:
            file = d['result']['item']['file']
        except:
            return {'error': 'No Item Playing'}
        data = d['result']['item']
        id = d['result']['item']['id']
        
        
        li = xbmcgui.ListItem(data['title'])  
        li.setInfo(self._type, data)
        
        if not xbmcvfs.exists(file):
            payload = {"jsonrpc": "2.0", "method": "Files.PrepareDownload", "params": {"path": file}, "id": "prepDownload"}
            r = requests.post(self._webserv + 'jsonrpc', data=json.dumps(payload))
            d = r.json()
            xbmc.log('kodi-mirror::download JSON: ' + json.dumps(d), xbmc.LOGDEBUG)
            stream = '%s%s' % (self._webserv, d['result']['details']['path'])
        else:
            stream = file

        return {'listItem': li, 'stream': stream, 'id': id}
    
    def _getSeek(self):
        payload = {"jsonrpc": "2.0", "method": "Player.GetProperties", "params": {"properties": ["time", "speed"], "playerid": self._pid}, "id": "getItem"}
        r = requests.post(self._webserv + 'jsonrpc', json=payload)
        d = r.json()
        xbmc.log('kodi-mirror::seek JSON: ' + json.dumps(d), xbmc.LOGDEBUG)
        seek = float(d['result']['time']['milliseconds']) / 1000
        seek += d['result']['time']['seconds']
        seek += d['result']['time']['minutes'] * 60
        seek += d['result']['time']['hours'] * 60 * 60
        return {"seek": seek, "speed": d['result']['speed']}
    
    def __del__(self):
        if self.socket is not None:
            self.socket.kill()
